### PR TITLE
ModuleList concatenation

### LINKED ISSUE #70441: [feature request] - ModuleList concatenation
## Issue description

Add concatenation for [`torch.nn.ModuleList`](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html).

## Code example

```python3
from torch import nn

a = [nn.Module()]
b = [nn.Module()]
c = a + b
print(c)

a = nn.ModuleList([nn.Module()])
b = nn.ModuleList([nn.Module()])
c = a + b # TypeError: unsupported operand type(s) for +: 'ModuleList' and 'ModuleList'
print(c)
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by jaketae**
Fixes https://github.com/pytorch/pytorch/issues/70441.

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks for the PR!

The lint failure is real though. Could you fix it?

**Comment by jaketae** [comment]
Thanks for the comment and review, @albanD!

The error comes from mypy, which complains that 

```
torch/nn/modules/container.py:214:5: error: Signatures of "__iadd__" and "__add__" are incompatible  [misc]
```

This error is raised because `__iadd__` accepts `Iterable[Module]`, whereas the newly implemented `__add__` expects `nn.ModuleList`, as this was the spirit of the original issue where this topic was raised. I can easily bypass this by changing `other` to be of type `Iterable[Module]` while maintaining the type check as show below.

```python
    def __add__(self, other: Iterable[Module]) -> 'ModuleList':
        if not isinstance(other, ModuleList):
            raise TypeError(...)
```

Part of me doesn't like this because of the inconsistency between the annotated type and what the functions actually wants. Alternatively, we can simply remove the type check and allow users to add `nn.ModuleList` and Python lists containing modules.

What do you think?

**Comment by jbschlosser** [comment]
> Alternatively, we can simply remove the type check and allow users to add nn.ModuleList and any Python iterable that contains modules.

My opinion is that we should go this route to match the signature of `__iadd__`. If we allow in-place addition with an iterable of modules, I don't see why we shouldn't also allow out-of-place addition of `ModuleList` + `Iterable[Module]`. @albanD you good with this?

**Comment by albanD** [comment]
I think adding an Iterable of Module or a ModuleList both work. I would expect both of these to be typed as `Union[Iterable[Module], ModuleList]` ? And make sure that both work fine via a test?

**Comment by jbschlosser** [comment]
> I think adding an Iterable of Module or a ModuleList both work. I would expect both of these to be typed as `Union[Iterable[Module], ModuleList]` ? And make sure that both work fine via a test?

The way I understand it is that `ModuleList` can be treated as a `Iterable[Module]`, so I don't think the Union is necessary. AFAICT it should work to have both `__add__` / `__iadd__` expect `Iterable[Module]` for the other arg and we get the ability to add two `ModuleLists` or a `ModuleList` + some other `Iterable[Module]`.

File: torch/nn/modules/container.py [hunk 1]
```diff
@@ -214,6 +214,21 @@ def __iter__(self) -> Iterator[Module]:
     def __iadd__(self, modules: Iterable[Module]) -> 'ModuleList':
         return self.extend(modules)
 
+    def __add__(self, other: 'ModuleList') -> 'ModuleList':
```
Comment (jaketae) [inline torch/nn/modules/container.py:217]:
```suggestion
    def __add__(self, other: Iterable[Module]) -> 'ModuleList':
```

**Review by jaketae — COMMENTED** [review]
@jbschlosser's explanation is also in line with my understanding. `Iterable[Module]` encompasses `ModuleList`, and the `__add__` signature should be identical to that of `__iadd__` to satisfy mypy. I'll update the signature and docs and see if relevant tests pass.

**Comment by jaketae** [comment]
@albanD @jbschlosser I'm having a bit of trouble understanding the test/build logs, such as [this](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.5-py3.7-test-distributed/382//console). Are the errors coming from this PR, and if so, what would be the next course of action I can take? Thanks in advance!

**Review by jbschlosser — APPROVED** [review]
LGTM - test failures look spurious

**Pull request closed without merging**